### PR TITLE
system76-scheduler: init at 2.0.1

### PR DESCRIPTION
**Pull request by cmm**
###### Description of changes

Package [System76-Scheduler](https://github.com/pop-os/system76-scheduler/).

This daemon:
- automatically tweaks CFS parameters for greater interactivity on mains power
- automatically sets certain processes to certain priorities according to rules that map executable names to desired priority
- automatically prioritizes foreground processes

There is nothing specific to System76 _hardware_ about this daemon, but the foreground prioritization part needs desktop shell support (since something must tell the daemon which processes are in foreground).  The Pop!_OS shell apparently supplies that functionality out of the box, and for Gnome there is the "System76 Scheduler" shell extension (which is in Nixpkgs and even works).

###### Things done

- Built on platform(s)
  - [v] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [v] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/module-list.nix [hunk 1]
```diff
@@ -408,6 +408,7 @@
   ./services/desktops/neard.nix
   ./services/desktops/profile-sync-daemon.nix
   ./services/desktops/system-config-printer.nix
+  ./services/desktops/system76-scheduler.nix
```
Comment (AndersonTorres) [inline nixos/modules/module-list.nix:411]: Put the module in a separate commit.

File: nixos/modules/services/desktops/system76-scheduler.nix [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+in {
+  options = {
+    services.system76-scheduler = {
+      enable = lib.mkEnableOption (lib.mdDoc "system76-scheduler");
+
+      package = lib.mkOption {
+        type = lib.types.package;
+        default = config.boot.kernelPackages.system76-scheduler;
+        defaultText = lib.literalExpression "config.boot.kernelPackages.system76-scheduler";
+        description = lib.mdDoc "Which System76-Scheduler package to use";
```
Comment (AndersonTorres) [inline nixos/modules/services/desktops/system76-scheduler.nix:23]: Such `description`s should end with period.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/system76-scheduler/default.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
+{ bcc, pkg-config, dbus, lib, rustPlatform, fetchFromGitHub, tree }:
+
+let
+  version = "25a45add4300eab47ceb332b4ec07e1e74e4baaf";
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/system76-scheduler/default.nix:4]: `version` should have the format `unstable-YYYY-MM-DD`.

Also, move this let inside the derivation.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/system76-scheduler/default.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
+{ bcc, pkg-config, dbus, lib, rustPlatform, fetchFromGitHub }:
+
+rustPlatform.buildRustPackage {
+  pname = "system76-scheduler";
+  version = "unstable-2022-10-05";
+  src = fetchFromGitHub {
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/system76-scheduler/default.nix:6]:
```suggestion

  src = fetchFromGitHub {
```

**Review by AndersonTorres — APPROVED** [review]
otherwise lgtm

File: lib/ron.nix [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+      else if RON.typeOf o == "some" then "Some(${toRON o.value})"
+      else if RON.typeOf o == "none" then "None"
+      else if RON.typeOf o == "tuple" then "${o.__RONTupleType or ""}(${concatStringsSep ", " (map toRON o.elements)})"
+      else if RON.typeOf o == "struct" then "${o.__RONStructType or ""}(${concatStringsSep ", "
+        (map (key: "${key}: ${toRON o.${key}}") (filter (n: n != "__RONStructType") (attrNames o)))})"
+      else if RON.typeOf o == "map" then "{${concatStringsSep ", " (map (m: "${toRON m.key}: ${toRON m.value}") o.mappings)}}"
+      else throw "Don't know how to serialize \"${RON.typeOf o}\" to RON"
+    else throw "Don't know how to serialize \"${typeOf o}\" to RON";
+in {
+  inherit RON fromRON toRON;
```
Comment (infinisil) [inline lib/ron.nix:283]: I don't think adding this kind of code to nixpkgs is a good idea. While it's neat that it works at all, Nix isn't meant for writing parsers and serializers. The code here is near impossible to maintain and comes with no tests. Furthermore it seems awful to use, since you need to know about all the RON specific Nix functions in addition to RON itself to create a valid configuration file.

File: nixos/modules/services/desktops/system76-scheduler.nix [hunk 6]
```diff
@@ -0,0 +1,102 @@
+{ config, lib, ... }:
+
+let
+  cfg = config.services.system76-scheduler;
+
+  readDefaultConfig = path:
+    lib.fromRON (builtins.unsafeDiscardStringContext (builtins.readFile "${cfg.package}/data/${path}"));
```
Comment (infinisil) [inline nixos/modules/services/desktops/system76-scheduler.nix:7]: This is using import-from-derivation, which is generally not allowed in nixpkgs. Since the result here is used to build the manual, it means that for every manual build by everybody, the system76 package is downloaded/built and the RON Nix parser run on the config files, which is definitely not okay.

Defaults like these just can't be exposed in the manual.

File: nixos/modules/services/desktops/system76-scheduler.nix [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        type = lib.types.listOf lib.types.str;
+        default = [];
+        description = lib.mdDoc "Programs that are left untouched; merged with `$(dirname $(realpath $(which system76-scheduler)))/../data/exceptions.ron`.";
+      };
+
+      assignments = lib.mkOption {
+        type = lib.RON.types.map (lib.types.listOf lib.types.str);
+        default = [];
+        description = lib.mdDoc "Assignments of priorities to programs; merged with `$(dirname $(realpath $(which system76-scheduler)))/../data/assignments.ron`.";
+      };
```
Comment (infinisil) [inline nixos/modules/services/desktops/system76-scheduler.nix:51]: How about just making this be `assignmentPaths` instead, allowing users to point to a RON file? Then we don't need to be involved in RON at all. In any case, options like these need to link to upstream documentation on how to specify configurations.

File: nixos/modules/services/desktops/system76-scheduler.nix [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        exceptions = default.exceptions;
+        assignments = default.assignments;
+      };
+    }
+
+    # add some obvious assignments and exceptions that are lacking
+    {
+      services.system76-scheduler = {
+        assignments = lib.RON.map [
+          { key = lib.RON.tuple [19 (lib.RON.identifier "Idle")];
```
Comment (infinisil) [inline nixos/modules/services/desktops/system76-scheduler.nix:68]: Again I don't think Nix should be involved in RON, this seems too hard to write in a Nix DSL

File: nixos/modules/services/desktops/system76-scheduler.nix [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+  config = lib.mkIf cfg.enable (lib.mkMerge [
+    # make sure stock defaults are merged with
+    {
+      services.system76-scheduler = {
+        exceptions = default.exceptions;
+        assignments = default.assignments;
+      };
+    }
+
+    # add some obvious assignments and exceptions that are lacking
```
Comment (infinisil) [inline nixos/modules/services/desktops/system76-scheduler.nix:64]: Should this be upstreamed? We shouldn't change the default configuration without a good reason

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/desktops/system76-scheduler.nix (hunk 9 above)
Comment (cmm) [reply nixos/modules/services/desktops/system76-scheduler.nix:64]: Yes, definitely.

File: nixos/modules/services/desktops/system76-scheduler.nix (hunk 8 above)
Comment (cmm) [reply nixos/modules/services/desktops/system76-scheduler.nix:68]: This could be way more user-friendly if Nix supported composite keys. :/

File: lib/ron.nix (hunk 5 above)
Comment (cmm) [reply lib/ron.nix:283]: Are you aware of, say, this: https://github.com/talyz/fromElisp/blob/master/default.nix ?
Nix is actually surprisingly handy at parsing -- it being a lazy functional language is enough of a boon and its shortcomings in string processing and matching are not enough of an impediment.

I fear this RON abomination is going to be very common soon, what with all the Rust hype, and in a few years' time `builtins.fromRON` is going to be a thing.

That all said, your points about maintenability and tests is well-taken. I'll redo this PR in a less ambitious way.

Comment (AndersonTorres) [reply lib/ron.nix:283]: > and in a few years' time builtins.fromRON is going to be a thing.

It needs a sanctioned grammar first, and well, Rust is far from being a stable language...

File: nixos/modules/services/desktops/system76-scheduler.nix (hunk 6 above)
Comment (infinisil) [reply nixos/modules/services/desktops/system76-scheduler.nix:7]: Ah you're right, I guess it only uses IFD if the module is enabled

File: lib/ron.nix [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+    len = stringLength string;
+
+    assemble = extract: init: plus: rs:
+      let
+        doit = rs:
+          if length rs == 0 then init
+          else (plus (extract (head rs)) (doit (tail rs)));
+      in
+        doit rs;
+    assembleString = assemble (_:_) "" (x: y: x + y);
```
Comment (chayleaf) [inline lib/ron.nix:111]: you can use lib.id and builtins.add here (of course this is just a small nitpick that doesn't affect much)

**Comment by cmm** [comment]
Converting this to draft because a proper new upstream release is imminent and a lot about this pull will have to be changed.  Also they are not using RON any more (I'm not sure the chosen alternative is materially better, but at least there's that)

**Comment by cmm** [comment]
packaging the newly-released version 2.0.0 now, no new parsers. :)

**Review by AndersonTorres — APPROVED** [review]
LGTM

**Comment by cmm** [comment]
force-pushed again because 2.0.1 is out, also cleaned up the module a bit.

File: nixos/modules/services/desktops/system76-scheduler.nix [hunk 11]
```diff
@@ -0,0 +1,310 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.system76-scheduler;
+
+  inherit (builtins) concatStringsSep map toString attrNames;
+  inherit (lib) boolToString types mkOption literalExpression mdDoc optional mkIf mkMerge;
+  inherit (types) nullOr listOf bool int ints float str enum;
+
+  latencyProfile = { latency
```
Comment (AndersonTorres) [inline nixos/modules/services/desktops/system76-scheduler.nix:10]: I don't know why, but this code is hopelessly ugly.

Comment (cmm) [reply nixos/modules/services/desktops/system76-scheduler.nix:10]: It definitely is, but why "hopelessly"?  Is it the overall approach, my lacking FP chops, the Nix language limitations, all of the above?

Comment (cmm) [reply nixos/modules/services/desktops/system76-scheduler.nix]: Any better now?

File: nixos/modules/services/desktops/system76-scheduler.nix [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+            "include descends=\"schedtool\""
+            "schedtool"
+          ]
+        '';
+        description = mdDoc "Processes that are left alone.";
+      };
+    };
+  };
+
+  config = {
```
Comment (misuzu) [inline nixos/modules/services/desktops/system76-scheduler.nix:229]: Looks like there is no way to disable this module.
```suggestion
  config = mkIf cfg.enable {
```

Comment (cmm) [reply nixos/modules/services/desktops/system76-scheduler.nix:229]: oh dear.  https://github.com/NixOS/nixpkgs/pull/229617
